File: Roadmap/14 - FECHAS/javascript/garos01.js
// Obteniendo la fecha actual
var fechaActual = new Date();

// Especificando la fecha de nacimiento
var fechaNacimiento = new Date("1994-09-11T13:40:03");

// Calculando la diferencia de tiempo en milisegundos
var diferenciaTiempo = fechaActual - fechaNacimiento;

// Convirtiendo la diferencia de tiempo en años
var añosTranscurridos = Math.floor(
  diferenciaTiempo / (1000 * 60 * 60 * 24 * 365)
);

// Imprimiendo el resultado
console.log(
  "Han transcurrido " +
    añosTranscurridos +
    " años desde mi fecha de nacimiento."
);

// Ejercicio extra

console.log(
  "1. Día, mes y año:",
  fechaNacimiento.getDate(),
  fechaNacimiento.getMonth() + 1,
  fechaNacimiento.getFullYear()
);
console.log(
  "2. Hora, minuto y segundo:",
  fechaNacimiento.getHours(),
  fechaNacimiento.getMinutes(),
  fechaNacimiento.getSeconds()
);
console.log(
  "3. Día de año:",
  Math.ceil(
    (fechaNacimiento - new Date(fechaNacimiento.getFullYear(), 0, 0)) / 86400000
  )
);
console.log("4. Día de la semana:", fechaNacimiento.getDay());
console.log("5. Nombre del mes:", obtenerNombreMes(fechaNacimiento.getMonth()));

// Función para obtener el nombre del mes
function obtenerNombreMes(numeroMes) {
  var meses = [
    "Enero",
    "Febrero",
    "Marzo",
    "Abril",
    "Mayo",
    "Junio",
    "Julio",
    "Agosto",
    "Septiembre",
    "Octubre",
    "Noviembre",
    "Diciembre",
  ];
  return meses[numeroMes];
}

// Formatos adicionales
console.log("6. Día del mes:", fechaNacimiento.getDate());
console.log("7. Año:", fechaNacimiento.getFullYear());
console.log("8. Hora:", fechaNacimiento.getHours());
console.log("9. Minuto:", fechaNacimiento.getMinutes());
console.log("10. Segundo:", fechaNacimiento.getSeconds());

File: Roadmap/01 - OPERADORES Y ESTRUCTURAS DE CONTROL/javascript/Roadmap/14 - FECHAS/python/garos01.py
from datetime import datetime

# Fecha actual
fecha_actual = datetime.now()

# Tu fecha de nacimiento (puedes modificar los valores)
fecha_nacimiento = datetime(year=1994, month=9, day=11, hour=13, minute=40, second=3)

# Calculando la diferencia de años
diferencia = fecha_actual.year - fecha_nacimiento.year

# Ajustando la diferencia si aún no ha pasado tu cumpleaños este año
if (fecha_actual.month, fecha_actual.day) < (
    fecha_nacimiento.month,
    fecha_nacimiento.day,
):
    diferencia -= 1

print("Han transcurrido {} años desde tu fecha de nacimiento.".format(diferencia))

# Ejercicio extra


formatos = [
    "%d/%m/%Y",  # Día, mes y año
    "%H:%M:%S",  # Hora, minuto y segundo
    "%j",  # Día de año
    "%A",  # Día de la semana
    "%B",  # Nombre del mes
    "%d de %B del %Y",  # Día de la semana completo
    "%m/%d/%y",  # Fecha en formato corto
    "%Y-%m-%d %H:%M:%S",  # Fecha y hora en formato ISO
    "%d/%m/%Y %H:%M:%S",  # Fecha y hora en formato personalizado
    "%I:%M %p",  # Hora en formato 12 horas con AM/PM
]

# Mostrar la fecha de nacimiento en los diferentes formatos
print("Fecha de nacimiento:")
for formato in formatos:
    print(fecha_nacimiento.strftime(formato))
